### ISSUE TITLE
Add a common location for python utility/tool files.

### ISSUE THREAD

**Issue by LGM-Doyle** [freeorion/freeorion#656]
There are three python portions of freeorion: universe_generation, turn_events and the AI.  Currently, they can only access files in their starting directory and below.  Consequently, utility/debugging code common to all three gets copied into each location.

There should be a common location for python utility classes, to avoid this code duplication.

**Comment by Cjkjvfnby** [comment]
This path can be added as `PYTHONPATH` at start of python.

**Comment by Vezzra** [comment]
Yes, absolutely agree, and I've already (long term) plans to implement something like that.

This includes moving the AI script directories to the `python` subfolder the `turn_events` and `universe_generation` script folders are already in - @freeorion/ai-team, is that ok with you? Then we can either place the "common" Python modules directly into the `python` folder, or create a dedicated subfolder (called "common" or "shared" or something like that). Any preferences/votes on that? Personally I'd prefer the latter option (a dedicated "common" subfolder).

**Comment by ghost** [comment]
I'll vote a common folder.  Arranging the scripts is actually pretty easy, but the Boost binding code does some strange things and will take time to fix.
[edit]
We don't need PYTHONPATH.  We can use `freeorion.__path__` to set up our imports.
It works like this:
1) Native freeorion module is just a stub.
2) Set `freeorion.__path__` to where our native extensions are and load the ones we want.
4) Set `freeorion.__path__` to $RESOURCEDIR/python
5) Profit.
[/edit]

**Comment by Cjkjvfnby** [comment]
This ticket is about pure python modules. 

``` diff
- import freeorion_utils
+ import commons.freeorion_utils
```

**Comment by LGM-Doyle** [comment]
All of the python started from within freeorion (generate universe, effects and AI) is already setup with the default/python directory as a common path.  I didn't realize this when I initially posted this issue.

Setting up a common location for code should be as simple as creating the directory under default/python.

**Comment by Cjkjvfnby** [comment]
> Setting up a common location for code should be as simple as creating the directory under default/python.

and add this directory as PYTHONPATH

**Comment by ghost** [comment]
Currently, all pure python modules are being loaded at global scope, leading to potential conflict with other packages.
Putting them all in default/python and adding that path as `freeorion.__path__` makes them all submodules of the `freeorion` module.

**Comment by ghost** [comment]
OK, I have a pure python setup that (mostly) passes pylint.  Modifying the native `freeorion` module to use the new structure will take more time, since that code is a bit of a maze.  (Read: doesn't actually work yet.  I probably missed something anyway)
WIP in Pull Request #707

**Comment by ghost** [comment]
There is a nasty circular import mess in the AI.  This should be considered a bug on its own.  Consider pull #707 on hold until it is fixed.

**Comment by Vezzra** [comment]
As adding a common location for python modules shared between AI, universe generation and turn event scripts will obviously be done together with other restructuring of the Python interface (won't make much sense otherwise), which means quite considerable changes, I suggest to flag this issue for post release.

Comments, opinions?

**Comment by Cjkjvfnby** [comment]
This issue not about moving python code.

It is about adding place for shared code.

At first iteration I expect just folder with single module "def test_shared(): print "it works"
that can be imported in all 3 modules.

**Comment by LGM-Doyle** [comment]
@Cjkjvfnby 
The PR `Add python logger support to AI` #669 already has a working shared code location.  The logger is in /python/utils/ and it works in AI and universe_generator.  I thought the issue had expanded into moving /AI into python/.

@spikethehobbitmage 
I did not run into a circular import problem.  I re-wrote parts of the CommonFramework.cpp and LoggingWrapper.cpp.  

However, I also did not account for the python global namespace import problems.  So choosing utils as the common directory name, may be problematic.

@Vezzra 
The following suggest may be perceived as self-serving.  So you could ignore it on those grounds.

Should we create either a post-0.4.6 branch or create the 0.4.6 branch now and allow master to accept post-0.4.6 commits?

The only issues marked for v0.4.6 that do not have associated PRs to fix them are.

#703 Mine damage in system with no mines                     ---------- dbenage working on it 
#693 Unable to load multiplayer saves.                             ---------- needs a multiplayer contributor.
#658 Compile error on Arch Linux 64bit with Boost 1.60   ---------- presumably trivial once any dev upgrades to Boost 1.60
#633 [AI] Research path needs to be checked and updated AI enhancement help wanted
#632 [AI] Weapon Upgrades aren't fully supported yet AI bug
#631 [AI] Tweak Drydock Repair logic AI enhancement
#630 [AI] Adaptations to new supply propagation required   ---------- Fix started by Cikjvfnby.
#547 Assign priorities to Industry and Research resource output effects groups     ---------- MatGB actively working on.

AI tuning is gating the release, specifically #633, #632 and #631.

The benefit is that non-AI PRs can be committed to master now and contributors are not asked in 2 months to rebase, clarify or re-work patches.   

The devs are disciplined enough to prioritize fixes for the release branch.

Produce the weekly builds from the pre-v0.4.6 branch to keep the non-compiling testers focused on the release version.  Or alternate.

We have a non-linear distributed version control system.  We should take advantage of that fact.

**Comment by Cjkjvfnby** [comment]
@LGM-Doyle, I miss that python directory is already in python path and can be used. So I can move shared code myself.

**Comment by Cjkjvfnby** [comment]
`default/python` folder added to PYTHONPATH between AI import and `generateOrder`.

Can someone fix just this?

To reproduce repalce `FreeOrionAI.generateOrders` line to 

``` python
print "On import", sys.path
@chat_on_error
@listener
def generateOrders():  # pylint: disable=invalid-name
    print "In generate orders", sys.path
```

PS I use repo folder as resource dir: `<resource-dir>f:\projects\FreeOrion\default</resource-dir>` but don't know if is change something

**Comment by ghost** [comment]
@Cjkjvfnby Fixing the PYTHONPATH mess is nontrivial.  I'm working on it, but it will require some major changes to how Freeorion uses python.
I use the repo folder, too.  It doesn't change anything.

**Comment by Vezzra** [comment]
Just adding `default/python` to PYTHONPATH is just a few lines of code, so not really that hard - technically. In practice I'm wary of the side effects this may cause, as the folders containing the universe generation and turn events scripts are subfolders of `default/python`, and I don't know if that might mess things up.

That's why I want to do all this properly, in one clean step: move the AI python scripts to `default/python/AI`, add `default/python/common` and use it as the location for Python modules shared between AI, universe generation and turn events.

Then, in a second step, clean up the "PYTHONPATH mess", as @spikethehobbitmage called it, by using a clear tree structure (more or less what @spikethehobbitmage suggests in #707).

As that doesn't sound like some trivial changes, I'd really prefer to postpone this until after the release. Unless there is a real need for these common Python modules already for 0.4.6. In that case I'd suggest to add the `default/python/common` folder, add it to PYTHONPATH and put shared modules in there. However, the question is: do we really need that?

**Comment by Vezzra** [comment]
@Cjkjvfnby: oh, sorry, I just now understood what you meant - `default/python` is already being added to `sys.path`, but _after_ initializing the C++/Python API modules (apparently that completely slipped my mind). Just pushed a fix for that, adding that path now happens _before_ initializing the API modules (968f09d01813d21012e893efc95669211266a81f).

Does that fix things properly for you? Can you use `default/python` for common Python modules now?

**Comment by Vezzra** [comment]
@spikethehobbitmage, I try to create the release branch as late as possible, usually shortly before we start to produce the release candidate builds. Non-linear distributed version control system is all nice and well, but in practice this means that as soon as the release branch has been created, the cherry picking fun commences.

Most bugs discovered need to be applied to both branches, unless it's a bug introduced only in master after the release branch has been created. Anything committed to master must be evaluated, if it should get ported over to the release branch. And so on. In short, it means a lot more work, especially for me, because as release manager I'm in charge of the release branch and have to keep track of what needs to go where.

And despite what you might expect concerning the discipline of our dev team, I know from experience that as soon as we have the release branch, people start to happily commit lots of stuff to master. I'm left to figure out what needs to go to the release branch too, which is obvious in most, but not all cases. Requiring me to nag people to clear these cases up.

Also, people often do _not_ cherry pick commits that should go to the release branch too themselves. Again requiring me to either do the cherry picking myself (which I do with simple changes) or nag people who committed the stuff to do it (in case of more complex changes and/or when conflicts arise).

Bottom line: no release branch until the latest moment possible :wink:

As for the test builds: once the release branch has been created, normal test builds are suspended. Only release candidate builds are provided, in case there is a longer gap between two release candidates I usually produce a weekly "pre-release" build based on the current release branch. Normal test builds resume after the release is out.

**Comment by Cjkjvfnby** [comment]
@Vezzra, will check it on next windows build.

Move AI to python and extract common code to special place are different features.  Move and change content will produce worst merges.

**Comment by Vezzra** [comment]
Well, of course I don't intend to do that in one humongous commit. :wink: What I have in mind is moving the AI scripts to `default/python/AI` and after that setting up `default/python/common` as shared location.

**Comment by Cjkjvfnby** [comment]
@Vezzra, order does not meter. I plan to extract some shared code before release (`print utils`, `autocomplete generation`, `config parsing`).

**Comment by ghost** [comment]
@Vezzra You are a saint for putting up with us.  Thank you.
Just moving the AI scripts to default/python/AI and updating the import path is trivial and works.  Cherry picking the second commit from #707 will do that.  Should I create a pull for just that?

**Comment by Vezzra** [comment]
> @Vezzra You are a saint for putting up with us. Thank you.

Well, you're welcome, but you're really not _that_ bad. :wink: I mean, you've to put up with me too after all, and I've not been around much lately, also kind of dragging my feet with some of the bigger PRs... 😔 

> Should I create a pull for just that?

If it's not too much hassle, yes.

**Comment by ghost** [comment]
@ Vezzra PR #714 ready.  It includes #713, which is also ready, so that should be applied first.  A simple pull is all that is needed for both.

**Comment by Cjkjvfnby** [comment]
@Vezzra lets merge it.

**Comment by Vezzra** [comment]
> PR #714 ready. It includes #713, which is also ready, so that should be applied first.

Just merging #714 effectively merged both PRs 😃
